contracts/TimelockVaultVotes.sol

```diff
@@ -0,0 +1,207 @@
+        // check that vesting doesn't already exit for the user, and if no - create one
+        if (accounts[to_user].cliff_timestamp == 0) {
+            accounts[to_user].cliff_timestamp = uint32(block.timestamp) + uint32(cliff_seconds);
+        }
+        if (accounts[to_user].vesting_ends_timestamp == 0) {
+            accounts[to_user].vesting_ends_timestamp = uint32(block.timestamp) + uint32(vesting_seconds);
+        }
+        addBalance(amount, to_user);
+    }
+
+    /**
+    * @notice Adds more tokens to the existing (possibly zero) vesting account. Doesn't change cliff or vesting period!
+    * @param amount Amount of token to be send to user under vesting, which will be deducted from msg.sender.
+    * @param to_user Beneficiary of the vesting account.
+    */
+    function addBalance(uint256 amount, address to_user) public {
+        // because we store amounts with loss of the precision we need to truck last 12 digits
+        LOCKED_TOKEN.safeTransferFrom(msg.sender, address(this), (amount / 1e12) * 1e12);
+        accounts[to_user].amount_under_vesting_div1e12 = accounts[to_user].amount_under_vesting_div1e12 + uint64(amount / 1e12);
+        _transferVotingUnits(address(0), msg.sender, (amount / 1e12) * 1e12);
+    }
+
+    /**
+     * @notice Checks amount of vested tokens minus already withdrawn.
+     * @return Returns amount of tokens the users can withdraw right now.
+     */
+    function availableToClaim(address user) public view returns (uint256) {
+        Account memory account = accounts[user];
+        uint256 t = block.timestamp;
+
+        // if the time is before the cliff - there's nothing vested yet
+        if (t < account.cliff_timestamp) return 0;
+
+        // if after the cliff
+        uint256 vested;
+        if (account.vesting_ends_timestamp < t) {
+            // if everything is vested
+            vested = uint256(account.amount_under_vesting_div1e12) * 1e12;
+        } else {
+            // otherwise the amount is proportional to the amount after the cliff before end of vesting
+            uint256 x = uint256(account.amount_under_vesting_div1e12) * 1e12;
+            uint256 y = (t - account.cliff_timestamp);
+            uint256 z = (uint256(account.vesting_ends_timestamp) - uint256(account.cliff_timestamp));
+            vested = x * y / z;
+        }
+        // the answer is how much is vested in total minute how much already withdrawn
+        return vested - uint256(account.amount_already_withdrawn_div1e12) * 1e12;
+    }
+
+    /**
+     * @notice User can claim their vested tokens.
+     */
+    function claim() external {
+        uint256 claim_amount = availableToClaim(msg.sender);
+        require(claim_amount > 0, "Can't claim 0 tokens");
+        accounts[msg.sender].amount_already_withdrawn_div1e12 = accounts[msg.sender].amount_already_withdrawn_div1e12 + uint64(claim_amount / 1e12);
+        _transferVotingUnits(msg.sender, address(0), (claim_amount / 1e12) * 1e12);
```

Comment:
could you please leave in the comments logic behind this code? 1) why does voting transfer is needed here? this is quivalent of mint function that "burns" veSuDAO token and give users SuDAO. 2) lock function is called by DAO only, before that token aren't delegated to anyone. 3) how does it solve problem for vote delegating for VeSuDAOv2? maybe write an example in the header comment of the contact what's the use-case step by step.

Reply:
Will do

Reply:
Added in commit 451b11758376e5f11abf0ba00ba338803a6c2f10